javalib/src/main/scala/java/util/regex/Matcher.scala

```diff
@@ -175,15 +175,259 @@ final class Matcher private[regex] (
+         case '\\' if s.length >= 2 =>
+           parse(s.substring(2), nextGroupIndex) match {
+             case (Node(Right(("","")), Left(regex)) +: tail, remaining, nextGroupIndex) =>
+               (Node(Right(("","")), Left("\\" + s(1) + regex)) +: tail, remaining, nextGroupIndex)
+             case (l: Seq[Node], remaining, nextGroupIndex) =>
+               (Seq(Node(Right(("","")), Left("\\" + s(1)))) ++ l, remaining, nextGroupIndex)
+           }
+         case '?' =>
+           parse(s.substring(1), nextGroupIndex) match {
+             case (l, r, i) =>
+               (Node(Right(("", s(0) + "")), Left("")) +: l, r, i)
+           }
+         case '+' | '*' => // greedy or not greedy
+           if (s.length >= 2 && s(1) == '?') {
+             parse(s.substring(2), nextGroupIndex) match {
+               case (l, r, i) =>
+                 (Node(Right(("", s(0) + "" + s(1))), Left("")) +: l, r, i)
+             }
+           } else {
+             parse(s.substring(1), nextGroupIndex) match {
+               case (l, r, i) =>
+                 (Node(Right(("", s(0) + "")), Left("")) +: l, r, i)
+             }
+           }
+         case '{' => // parse until end of occurrence
+           val (inside, remaining) = parseUntilEndBrace(s.substring(1))
+           parse(remaining, nextGroupIndex) match {
+             case (l, r, i) =>
+                (Node(Right(("", "{" + inside)), Left("")) +: l, r, i)
+           }
+         case '[' =>
+            val (inside, remaining) = parseUntilEndSquareBracket(s.substring(1))
+           parse(remaining, nextGroupIndex) match {
+             case (l, r, i) =>
+                (Node(Right(("", "")), Left("[" + inside)) +: l, r, i)
+           }
+         case e => 
+           parse(s.substring(1), nextGroupIndex) match {
+             case (Node(Right(("","")), Left(regex)) +: tail, remaining, nextGroupIndex) =>
+               (Node(Right(("","")), Left(e + regex)) +: tail, remaining, nextGroupIndex)
+             case (l: Seq[Node], remaining, nextGroupIndex) =>
+               (Seq(Node(Right(("","")), Left("" + e))) ++ l, remaining, nextGroupIndex)
+           }
+         }
+         res match {
+           case (Node(Right((left, right)), m) +:
+                 Node(Right(("", modifier)), Left("")) +: tail, r, i)
+                 if modifier == "?" || modifier == "*" || modifier == "+" || modifier == "*?"
+                  || modifier == "+?" || modifier.startsWith("{") =>
+                  
+                (Node(Right((left, right + modifier)), m) +: tail, r, i)
+           case ((n@Node(Left(_), m)) +:
+                 Node(Right(("", modifier)), Left("")) +: tail, r, i)
+                 if modifier == "?" || modifier == "*" || modifier == "+" || modifier == "*?"
+                  || modifier == "+?" || modifier.startsWith("{") =>
+                 (Node(Right(("", modifier)), Right(Seq(n))) +: tail, r, i)
+           case _ => res
```

Comment:
What does this do? There's got to be a nicer way to write this stuff.

Reply:
If I parse a group (A) and then a modifier like the one mentioned, then  I rewrite the nodes by adding the modifiers as a right addition is the group was not matching, and wrap in another node if the group was already matching.

To be nicer, I pattern matched on the second node of the list, and then on the first.